lib/tbot/spiffe/workloadattest/kubernetes.go

```diff
@@ -0,0 +1,418 @@
+	// valid certificate signed by the cluster CA. This is more common than
+	// you'd think.
+	SkipVerify bool `yaml:"skip_verify,omitempty"`
+	// Anonymous is used to indicate that no authentication should be used
+	// when connecting to the secure Kubelet API. If set, TokenPath will be
+	// ignored.
+	Anonymous bool `yaml:"anonymous,omitempty"`
+}
+
+func (c KubeletClientConfig) CheckAndSetDefaults() error {
+	if c.ReadOnlyPort != 0 && c.SecurePort != 0 {
+		return trace.BadParameter("readOnlyPort and securePort are mutually exclusive")
+	}
+	return nil
+}
+
+// kubeletClient is a HTTP client for the Kubelet API
+type kubeletClient struct {
+	cfg    KubeletClientConfig
+	getEnv func(string) string
+}
+
+func newKubeletClient(cfg KubeletClientConfig) *kubeletClient {
+	return &kubeletClient{
+		cfg:    cfg,
+		getEnv: os.Getenv,
+	}
+}
+
+type roundTripperFn func(req *http.Request) (*http.Response, error)
+
+func (f roundTripperFn) RoundTrip(req *http.Request) (*http.Response, error) {
+	return f(req)
+}
+
+func (c *kubeletClient) httpClient() (url.URL, *http.Client, error) {
+	host := c.getEnv(nodeNameEnv)
+
+	if c.cfg.ReadOnlyPort != 0 {
+		return url.URL{
+			Scheme: "http",
+			Host:   net.JoinHostPort(host, strconv.Itoa(c.cfg.ReadOnlyPort)),
+		}, &http.Client{}, nil
+	}
+
+	port := cmp.Or(c.cfg.SecurePort, defaultSecurePort)
+
+	transport := &http.Transport{
+		TLSClientConfig: &tls.Config{},
+	}
+
+	switch {
+	case c.cfg.SkipVerify:
+		transport.TLSClientConfig.InsecureSkipVerify = true
+	default:
+		caPath := cmp.Or(c.cfg.CAPath, defaultCAPath)
+		certPool := x509.NewCertPool()
+		caPEM, err := os.ReadFile(caPath)
+		if err != nil {
+			return url.URL{}, nil, trace.Wrap(err, "reading CA file %q", caPath)
+		}
+		if !certPool.AppendCertsFromPEM(caPEM) {
+			return url.URL{}, nil, trace.BadParameter("failed to append CA cert from %q", caPath)
+		}
+		transport.TLSClientConfig.RootCAs = certPool
+	}
+
+	client := &http.Client{
+		Transport: transport,
+	}
+
+	switch {
+	case c.cfg.Anonymous:
+	// Nothing to do
+	case c.cfg.TokenPath != "":
+		fallthrough
+	default:
+		tokenPath := cmp.Or(c.cfg.TokenPath, defaultServiceAccountTokenPath)
+		token, err := os.ReadFile(tokenPath)
+		if err != nil {
+			return url.URL{}, nil, trace.Wrap(err, "reading token file %q", tokenPath)
+		}
+		client.Transport = roundTripperFn(func(req *http.Request) (*http.Response, error) {
+			req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+			return transport.RoundTrip(req)
+		})
+	}
+
+	return url.URL{
+		Scheme: "https",
+		Host:   net.JoinHostPort(host, strconv.Itoa(port)),
+	}, client, nil
+}
+
+func (c *kubeletClient) ListAllPods(ctx context.Context) (*v1.PodList, error) {
+	reqUrl, client, err := c.httpClient()
+	if err != nil {
+		return nil, trace.Wrap(err, "creating HTTP client")
+	}
+	reqUrl.Path = "/pods"
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, reqUrl.String(), nil)
```

Comment:
I wonder if this would benefit from a timeout or something, I accidentally had `hostNetwork: false` set (i.e. the default value) and it just hangs forever in a very difficult to debug way.

Reply:
Agreed !